### PR TITLE
Join public class

### PR DESCRIPTION
**Pull request by aaronmsegal**
Changes to allow users to join public classes listed in the public class list. The following changes were made

1. app.py now has a joinClass method at the '/user/classes/' POST endpoint
2. class_ctrl.py has some added logging
3. In user_ctrl.py, users are now created without classCode lists. This is now added as a set the first time a user joins a class.
4. user_ctrl.py now has a joinClass  method. When called the class's code is added to the user's classCode set and the user's email is added to the class's student set .
5. Backend testing now uses mock to avoid trying to send an email, but we can use this for alot of other things as well
6. Added a test method fro the new prepForJsonDump method in ControllerResponse
7. Modified test_user_ctrl.py to setup both user and class tables in the same way that  test_class_ctrl.py does
8. Added a test method for user_ctrl.py's joinClass  method
9. ResponseCreation.py now has a prepForJsonDump method, which is called every time a response's payload is added to. This method confirms that the payload is ready to be converted to json by recursively replacing all sets and tuples with lists.
10. browse.component.ts can now subscribe to and handle the user service call to joinClass
11. Join class button now actually calls methods to join class
12. user.service.ts now has a joinClass method to call the backend's '/user/classes/' POST endpoint

To test: 
register and activate a new student. If there is already a public class listed, click the join class button to join. It should be removed from the public class list and added to the student's class list.  If there is not a public class use the admin account to add a new class, then switch back to the student and join the class.

### CODE REVIEW

File: Backend/mentii/class_ctrl.py [hunk 1]
```diff
@@ -116,7 +117,9 @@ def getClassCodesFromUser(dynamoDBInstance, email=None):
     request = {"Key" : {"email": email}, "ProjectionExpression": "classCodes"}
     res = dbUtils.getItem(request, usersTable)
     #Get the class codes for the user.
-    if res is not None and 'Item' in res:
+    if res is None or 'Item' not in res or 'classCodes' not in res['Item']:
```
Comment (RyanYoung25) [inline Backend/mentii/class_ctrl.py]: Nice catch on the extra, better, error checking!

File: Backend/tests/test_user_ctrl.py [hunk 2]
```diff
@@ -257,6 +269,60 @@ def test_getRole_fail(self):
...
+      allClassCodes.add(code)
+      joinData = { 'code' : code }
+      res = usr.joinClass(joinData, dynamodb, joinEmail)
+      self.assertFalse(res.hasErrors())
+      self.assertEqual(res.payload['code'], code)
+
+    joined = classCtrl.getClassCodesFromUser(dynamodb, joinEmail)
+    missing = joined - allClassCodes
+    #confirm none missing
+    self.assertFalse(missing)
```
Comment (RyanYoung25) [inline Backend/tests/test_user_ctrl.py]: So an empty set is the same as false? Not something that you need to change but it just wasn't immediately clear and I wanted to make sure that the test makes sense.

File: Backend/mentii/class_ctrl.py [hunk 3]
```diff
@@ -116,7 +117,9 @@ def getClassCodesFromUser(dynamoDBInstance, email=None):
...
+      MentiiLogging.getLogger().error('Unable to get user data in getClassCodesFromUser')
+    else:
       classCodes = res['Item']['classCodes']
   return classCodes
```
Comment (RyanYoung25) [inline Backend/mentii/class_ctrl.py]: One thing to note, we initialize classCodes as a list at the top of the method and then return what I assume is a set from the res['Item'] dictionary. We should probably initialize classCodes for consistency and to know that this function is going to return a set no matter what.

File: Backend/mentii/user_ctrl.py [hunk 4]
```diff
@@ -252,3 +251,51 @@ def getRole(userEmail, dynamoDBInstance):
       userRole = res['Item']['userRole']
 
   return userRole
+
+def joinClass(jsonData, dynamoDBInstance, email=None):
```
Comment (jonathanmcdaniel) [inline Backend/mentii/user_ctrl.py]: The complexity of this function is pretty high. Do you think you could break out some of these if/else into other functions that you can just call?

File: Backend/app.py [hunk 5]
```diff
@@ -170,16 +170,25 @@ def class_list():
     status = 400
   return ResponseCreation.createResponse(res, status)
 
+@app.route('/user/classes/', methods=['POST'])
```
Comment (aaronmsegal) [inline Backend/app.py]: this is caught by:
```
@app.route('/user/classes/', methods=['GET', 'OPTIONS'])
@auth.login_required		 @auth.login_required
def class_list():		 def class_list(): 
```
at line 161

File: Backend/mentii/class_ctrl.py (hunk 3 above)
Comment (aaronmsegal) [inline Backend/mentii/class_ctrl.py]: This is now changed

File: Backend/mentii/user_ctrl.py (hunk 4 above)
Comment (aaronmsegal) [inline Backend/mentii/user_ctrl.py]: I split adding class to user and user to class out into their own methods.

File: Backend/tests/test_controllerResponse.py [hunk 6]
```diff
@@ -46,5 +46,38 @@ def test_hasErrors(self):
...
+    cleaned = self.response.prepForJsonDump(testList)
+    self.assertTrue(isinstance(cleaned, list))
+    self.assertTrue(len(cleaned), 4)
+    self.assertTrue(isinstance(cleaned[2], list))
+    self.assertTrue(isinstance(cleaned[3], list))
+
+    testDict['list'] = testList
+    cleaned = self.response.prepForJsonDump(testDict)
+    self.assertTrue(isinstance(cleaned, dict))
+    self.assertTrue(isinstance(cleaned['list'], list))
```
Comment (aaronmsegal) [inline Backend/tests/test_controllerResponse.py]: This one is ok

File: Backend/tests/test_user_ctrl.py [hunk 7]
```diff
@@ -109,11 +123,9 @@ def test_addUserAndSendEmail(self):
     password = "password8"
 
     activationId = ""
```
Comment (aaronmsegal) [inline Backend/tests/test_user_ctrl.py]: Not part of this pull request

File: Backend/tests/test_user_ctrl.py [hunk 8]
```diff
@@ -257,6 +269,60 @@ def test_getRole_fail(self):
     userRole = usr.getRole('test85@mentii.me', dynamodb)
     self.assertIsNone(userRole)
 
+  def test_joinClass(self):
+    print 'Running test_joinClass'
```
Comment (aaronmsegal) [inline Backend/tests/test_user_ctrl.py]: yep  not python 3

File: Backend/tests/test_user_ctrl.py [hunk 9]
```diff
@@ -305,7 +305,7 @@ def test_joinClass(self):
 
     joined = classCtrl.getClassCodesFromUser(dynamodb, joinEmail)
     missing = joined - allClassCodes
-    #confirm none missing
+    #confirm the empty set, meing that all classes have been joined
```
Comment (RyanYoung25) [inline Backend/tests/test_user_ctrl.py]: the set is empty, meaning that all classes have been joined*
